File: app/src/main/java/xtr/keymapper/activity/ImportExportActivity.java
package xtr.keymapper.activity;

import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.content.res.AppCompatResources;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.card.MaterialCardView;

import java.io.BufferedReader;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;
import java.util.zip.ZipOutputStream;

import xtr.keymapper.R;
import xtr.keymapper.databinding.ActivityImportExportBinding;
import xtr.keymapper.databinding.ProfileRowItem2Binding;
import xtr.keymapper.keymap.KeymapProfile;
import xtr.keymapper.keymap.KeymapProfiles;

public class ImportExportActivity extends AppCompatActivity {
    private static final int WRITE_REQUEST_CODE = 101;
    private static final int READ_REQUEST_CODE = 102;
    private ByteArrayOutputStream byteArrayOutputStream = null;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ActivityImportExportBinding binding = ActivityImportExportBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        ProfilesViewAdapter profilesViewAdapter = new ProfilesViewAdapter(this);
        binding.profiles.setAdapter(profilesViewAdapter);

        binding.selectAllButton.setOnClickListener(v -> {
            for (int i = 0; i < profilesViewAdapter.getItemCount(); i++)
                ((ProfilesViewAdapter.ViewHolder) binding.profiles.findViewHolderForAdapterPosition(i)).binding.card.setChecked(true);
        });
        binding.exportButton.setOnClickListener(this::exportProfiles);
        binding.importButton.setOnClickListener(this::openProfile);
    }

    private void openProfile(View view) {
        Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
        intent.setType("application/zip");
        intent.addCategory(Intent.CATEGORY_OPENABLE);
        startActivityForResult(intent, READ_REQUEST_CODE);
    }


    private void exportProfiles(View view) {
        RecyclerView recyclerView = findViewById(R.id.profiles);
        ArrayList<String> profileNames = new ArrayList<>();
        for (int i = 0; i < recyclerView.getAdapter().getItemCount(); i++) {
            MaterialCardView card = ((ProfilesViewAdapter.ViewHolder) recyclerView.findViewHolderForAdapterPosition(i)).binding.card;
            if (card.isChecked()) {
                profileNames.add(((ProfilesViewAdapter)recyclerView.getAdapter()).recyclerDataArrayList.get(i).name);
            }
        }

        KeymapProfiles keymapProfiles = new KeymapProfiles(this);
        try {
            byteArrayOutputStream = new ByteArrayOutputStream();
            ZipOutputStream zipOutputStream = new ZipOutputStream(byteArrayOutputStream);
            profileNames.forEach(profileName -> {
                ZipEntry zipEntry = new ZipEntry(profileName);
                try {
                    zipOutputStream.putNextEntry(zipEntry);
                    keymapProfiles.sharedPref.getStringSet(profileName, new HashSet<>()).forEach(s -> {
                        try {
                            zipOutputStream.write(s.getBytes());
                            zipOutputStream.write("\n".getBytes());
                        } catch (IOException e) {
                            throw new RuntimeException(e);
                        }
                    });
                    zipOutputStream.closeEntry();
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            });
            zipOutputStream.close();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        Intent intent = new Intent(Intent.ACTION_CREATE_DOCUMENT);
        intent.putExtra(Intent.EXTRA_TITLE, "backup.zip");
        intent.setType("application/zip");
        intent.addCategory(Intent.CATEGORY_OPENABLE);
        startActivityForResult(intent, WRITE_REQUEST_CODE);
    }

    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == WRITE_REQUEST_CODE && resultCode == RESULT_OK) {
                    if (data != null
                            && data.getData() != null
                                && byteArrayOutputStream != null) {
                        try ( OutputStream outputStream =
                                      getContentResolver().openOutputStream(data.getData())) {
                            outputStream.write(byteArrayOutputStream.toByteArray());
                            outputStream.close();
                            byteArrayOutputStream = null;
                        } catch (IOException e) {
                            throw new RuntimeException(e);
                        }
                    }

        } else if (requestCode == READ_REQUEST_CODE && resultCode == RESULT_OK) {
            if (data != null
                    && data.getData() != null) {
                importProfiles(data.getData());
            }
        }
    }

    private void importProfiles(Uri dataUri) {
        try (InputStream inputStream =
                     getContentResolver().openInputStream(dataUri)) {
            ZipInputStream zipInputStream = new ZipInputStream(inputStream);
            ZipEntry zipEntry;
            while ((zipEntry = zipInputStream.getNextEntry()) != null) {
                String profileName = zipEntry.getName();
                Set<String> stringSet = new HashSet<>();
                BufferedReader reader = new BufferedReader(
                        new InputStreamReader(zipInputStream));
                String line;
                while ((line = reader.readLine()) != null) {
                    stringSet.add(line);
                }
                new KeymapProfiles(this).sharedPref
                        .edit()
                        .putStringSet(profileName, stringSet)
                        .apply();
                zipInputStream.closeEntry();
            }
            zipInputStream.close();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        // Refresh RecyclerView

        ProfilesViewAdapter profilesViewAdapter = new ProfilesViewAdapter(this);
        RecyclerView recyclerView = findViewById(R.id.profiles);
        recyclerView.setAdapter(profilesViewAdapter);
    }

    static class ProfilesViewAdapter extends RecyclerView.Adapter<ProfilesViewAdapter.ViewHolder> {

        private final ArrayList<RecyclerData> recyclerDataArrayList = new ArrayList<>();

        /**
         * Provide a reference to the type of views used
         */
        public static class ViewHolder extends RecyclerView.ViewHolder {

            private final ProfileRowItem2Binding binding;

            public ViewHolder(ProfileRowItem2Binding binding) {
                super(binding.getRoot());
                this.binding = binding;
            }
        }

        /**
         * Initialize the dataset of the Adapter.
         */
        public ProfilesViewAdapter(Context context) {
            if (context == null) return;
            KeymapProfiles keymapProfiles = new KeymapProfiles(context);

            new KeymapProfiles(context).getAllProfiles().forEach((profileName, profile) -> {
                if(profileName != null)
                    recyclerDataArrayList.add(new RecyclerData(profile, context, profileName));
                else keymapProfiles.deleteProfile(null);
            });
        }
        // Create new views (invoked by the layout manager)
        @NonNull
        @Override
        public ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {
            // Create a new view
            ProfileRowItem2Binding itemBinding = ProfileRowItem2Binding.inflate(LayoutInflater.from(viewGroup.getContext()), viewGroup, false);
            return new ViewHolder(itemBinding);
        }

        // Replace the contents of a view (invoked by the layout manager)
        @Override
        public void onBindViewHolder(ViewHolder viewHolder, final int position) {
            // Get element from dataset at this position and set the contents of the view
            RecyclerData recyclerData = recyclerDataArrayList.get(position);
            viewHolder.binding.appIcon.setImageDrawable(recyclerData.icon);
            viewHolder.binding.profileName.setText(recyclerData.name);
            viewHolder.binding.profileText.setText(recyclerData.description);
            viewHolder.binding.card.setOnClickListener(v -> viewHolder.binding.card.setChecked(!viewHolder.binding.card.isChecked()));
        }

        // Return the size of dataset (invoked by the layout manager)
        @Override
        public int getItemCount() {
            return recyclerDataArrayList.size();
        }

        private static class RecyclerData {
            public RecyclerData(KeymapProfile profile, Context context, String name) {
                this.name = name;
                this.description = new KeymapProfiles(context).sharedPref.getStringSet(name, new HashSet<>()).toString();
                try {
                    this.icon = context.getPackageManager().getApplicationIcon(profile.packageName);
                } catch (PackageManager.NameNotFoundException e) {
                    this.icon = AppCompatResources.getDrawable(context, R.mipmap.ic_launcher_foreground);
                }
            }

            String description;
            String name;
            Drawable icon;
        }
    }

}

File: app/src/main/java/xtr/keymapper/activity/ImportExportActivity.kt
package xtr.keymapper.activity

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.drawable.Drawable
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.content.res.AppCompatResources
import androidx.core.content.edit
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.snackbar.Snackbar
import xtr.keymapper.R
import xtr.keymapper.databinding.ActivityImportExportBinding
import xtr.keymapper.databinding.ProfileRowItem2Binding
import xtr.keymapper.keymap.KeymapProfile
import xtr.keymapper.keymap.KeymapProfiles
import java.io.*
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream
import java.util.zip.ZipOutputStream

class ImportExportActivity : AppCompatActivity() {
    private var byteArrayOutputStream: ByteArrayOutputStream? = null
    private lateinit var binding: ActivityImportExportBinding

    // SoufianoDev: Modern Activity Result Launchers For API 30+
    private lateinit var createDocumentLauncher: ActivityResultLauncher<Intent>
    private lateinit var openDocumentLauncher: ActivityResultLauncher<Intent>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityImportExportBinding.inflate(layoutInflater)
        setContentView(binding.root)

        // SoufianoDev: Initialize Launchers Conditionally Based On OS Version
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            initModernLaunchers()
        }

        setupRecyclerView(false)

        binding.selectAllButton.setOnClickListener { setupRecyclerView(true) }
        binding.importButton.setOnClickListener { openZipFile() }
    }

    private fun initModernLaunchers() {
        // SoufianoDev: Register Callback For Saving Documents (Export)
        createDocumentLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
            if (result.resultCode == RESULT_OK) {
                val uri = result.data?.data
                val bytes = byteArrayOutputStream?.toByteArray()
                // SoufianoDev: Smart Cast With Null Check
                if (uri != null && bytes != null) writeBytesToUri(uri, bytes)
            }
        }

        // SoufianoDev: Register Callback For Picking Documents (Import)
        openDocumentLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
            if (result.resultCode == RESULT_OK) {
                result.data?.data?.let { importProfiles(it) }
            }
        }
    }

    private fun setupRecyclerView(allChecked: Boolean) {
        binding.profiles.adapter = ProfilesViewAdapter(this, allChecked)
    }

    private fun openZipFile() {
        val intent = Intent(Intent.ACTION_OPEN_DOCUMENT).apply {
            type = "application/zip"
            addCategory(Intent.CATEGORY_OPENABLE)
        }

        // SoufianoDev: Version-Based Branching For Activity Results
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            openDocumentLauncher.launch(intent)
        } else {
            @Suppress("DEPRECATION")
            startActivityForResult(intent, READ_REQUEST_CODE)
        }
    }

    private fun exportProfiles(profileNames: List<String>) {
        if (profileNames.isEmpty()) return
        val keymapProfiles = KeymapProfiles(this)

        try {
            val baos = ByteArrayOutputStream()
            // SoufianoDev: Use Function For Auto Close
            ZipOutputStream(baos).use { z : ZipOutputStream ->
                profileNames.forEach { profileName ->
                    val zipEntry = ZipEntry(profileName)
                    z.putNextEntry(zipEntry)
                    keymapProfiles.sharedPref.getStringSet(profileName, emptySet())?.forEach { s ->
                        z.write(s.toByteArray())
                        z.write("\n".toByteArray())
                    }
                    z.closeEntry()
                }
            }
            byteArrayOutputStream = baos

            val intent = Intent(Intent.ACTION_CREATE_DOCUMENT).apply {
                // SoufianoDev: Use Localized Default Filename From Strings.xml
                putExtra(Intent.EXTRA_TITLE, getString(R.string.default_backup_name))
                type = "application/zip"
                addCategory(Intent.CATEGORY_OPENABLE)
            }

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                createDocumentLauncher.launch(intent)
            } else {
                @Suppress("DEPRECATION")
                startActivityForResult(intent, WRITE_REQUEST_CODE)
            }

        } catch (e: IOException) {
            Log.e("ImportExport", "Export process failed", e)
            // SoufianoDev: Snackbar With Localized Strings (Retry Action)
            Snackbar.make(binding.root, getString(R.string.error_export_failed), Snackbar.LENGTH_LONG)
                .setAction(getString(R.string.action_retry)) { exportProfiles(profileNames) }
                .show()
        }
    }

    @Deprecated("Deprecated in Java")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        @Suppress("DEPRECATION")
        super.onActivityResult(requestCode, resultCode, data)

        // SoufianoDev: Legacy Handler For Devices Below Android 11
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R && resultCode == RESULT_OK && data != null) {
            when (requestCode) {
                WRITE_REQUEST_CODE -> {
                    val uri = data.data
                    val bytes = byteArrayOutputStream?.toByteArray()
                    // SoufianoDev: Smart Cast With Null Check
                    if (uri != null && bytes != null) writeBytesToUri(uri, bytes)
                }
                READ_REQUEST_CODE -> {
                    data.data?.let { importProfiles(it) }
                }
            }
        }
    }

    // SoufianoDev: Helper Method For Atomic Write
    private fun writeBytesToUri(uri: Uri, bytes: ByteArray) {
        try {
            contentResolver.openOutputStream(uri)?.use { os ->
                os.write(bytes)
            }
            byteArrayOutputStream = null
            // SoufianoDev: Localized Success Message
            Snackbar.make(binding.root, getString(R.string.msg_backup_saved), Snackbar.LENGTH_SHORT).show()
        } catch (e: IOException) {
            Log.e("ImportExport", "Failed to write bytes to URI", e)
            // SoufianoDev: Retry Writing From SnackBar Action (Localized UI)
            Snackbar.make(binding.root, getString(R.string.error_save_failed), Snackbar.LENGTH_LONG)
                .setAction(getString(R.string.action_retry)) { writeBytesToUri(uri, bytes) }
                .show()
        }
    }

    @SuppressLint("UseKtx")
    private fun importProfiles(dataUri: Uri) {
        try {
            contentResolver.openInputStream(dataUri)?.use { inputStream ->
                ZipInputStream(inputStream).use { zis ->
                    var zipEntry: ZipEntry? = zis.nextEntry
                    val profilesManager = KeymapProfiles(this)

                    while (zipEntry != null) {
                        // SoufianoDev: Safe Property Access Instead Of Non-Null Assertion
                        val profileName = zipEntry.name
                        val stringSet = mutableSetOf<String>()
                        // SoufianoDev: Reader Without Closing Inner Stream
                        val reader = BufferedReader(InputStreamReader(zis))

                        var line: String? = reader.readLine()
                        while (line != null) {
                            stringSet.add(line)
                            line = reader.readLine()
                        }

                        // SoufianoDev: Use KTX Extension For Cleaner SharedPreferences Editing
                        profilesManager.sharedPref.edit {
                            putStringSet(profileName, stringSet)
                        }

                        zis.closeEntry()
                        zipEntry = zis.nextEntry
                    }
                }
            }
            // SoufianoDev: Localized Import Success Notification
            Snackbar.make(binding.root, getString(R.string.msg_profiles_imported), Snackbar.LENGTH_SHORT).show()
        } catch (e: IOException) {
            Log.e("ImportExport", "Import failed from URI", e)
            // SoufianoDev: Localized Fail/Retry Logic
            Snackbar.make(binding.root, getString(R.string.error_import_failed), Snackbar.LENGTH_LONG)
                .setAction(getString(R.string.action_retry)) { importProfiles(dataUri) }
                .show()
        }
        setupRecyclerView(false)
    }

    internal inner class ProfilesViewAdapter(context: Context, private val allCardsChecked: Boolean) :
        RecyclerView.Adapter<ProfilesViewAdapter.ViewHolder>() {

        private val items = mutableListOf<RecyclerData>()
        private val selectedProfileNames = mutableListOf<String>()

        inner class ViewHolder(val itemBinding: ProfileRowItem2Binding) : RecyclerView.ViewHolder(itemBinding.root)

        init {
            val profilesManager = KeymapProfiles(context)
            profilesManager.allProfiles.forEach { (name, profile) ->
                // SoufianoDev: Safe Unwrapping Of Profile Name
                if (name != null) {
                    items.add(RecyclerData(profile, context, name))
                    if (allCardsChecked) selectedProfileNames.add(name)
                } else {
                    profilesManager.deleteProfile(null)
                }
            }
            binding.exportButton.setOnClickListener { exportProfiles(selectedProfileNames) }
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            val inflater = LayoutInflater.from(parent.context)
            return ViewHolder(ProfileRowItem2Binding.inflate(inflater, parent, false))
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            val data = items[position]
            // SoufianoDev: Scope Function To Group
            with(holder.itemBinding) {
                appIcon.setImageDrawable(data.icon)
                profileName.text = data.name
                profileText.text = data.description
                card.isChecked = allCardsChecked || selectedProfileNames.contains(data.name)

                card.setOnClickListener {
                    card.isChecked = !card.isChecked
                    if (card.isChecked) {
                        if (!selectedProfileNames.contains(data.name)) selectedProfileNames.add(data.name)
                    } else {
                        selectedProfileNames.remove(data.name)
                    }
                }
            }
        }

        override fun getItemCount() = items.size

        private inner class RecyclerData(profile: KeymapProfile, context: Context, val name: String) {
            val description: String = KeymapProfiles(context).sharedPref
                .getStringSet(name, emptySet()).toString()

            // SoufianoDev: Modern Resource Access Pattern
            val icon: Drawable? = try {
                context.packageManager.getApplicationIcon(profile.packageName)
            } catch (_: PackageManager.NameNotFoundException) {
                AppCompatResources.getDrawable(context, R.mipmap.ic_launcher_foreground)
            }
        }
    }

    companion object {
        private const val WRITE_REQUEST_CODE = 101
        private const val READ_REQUEST_CODE = 102
    }
}